BBC: Blead Breaks Sereal::Encoder

This is a bug report for perl from "Carlos Guevara" <carlos@carlosguevara.com>,
generated with the help of perlbug 1.43 running under perl 5.41.8.


-----------------------------------------------------------------
<!--[Please describe your issue here]-->

BBC: Blead Breaks Sereal::Encoder

Please see http://fast-matrix.cpantesters.org/?dist=Sereal::Encoder

<!--[Please do not change anything below this line]-->
<!------------------------------------------------------------------- -->


---
**Flags**
- category=core
- severity=low
---
**Perl configuration**
```
Site configuration information for perl 5.41.8:

Configured by cpan at Sat Jan 11 02:10:58 EST 2025.

Summary of my perl5 (revision 5 version 41 subversion 8) configuration:
  Commit id: 3a16cb5fb8114b7773a9c488c78684bffd787b62
  Platform:
    osname=linux
    osvers=5.14.0-503.21.1.el9_5.x86_64
    archname=x86_64-linux
    uname='linux cjg-rhel9 5.14.0-503.21.1.el9_5.x86_64 #1 smp preempt_dynamic thu dec 19 09:37:00 est 2024 x86_64 x86_64 x86_64 gnulinux '
    config_args='-des -Dprefix=/home/cpan/bin/perl -Dscriptdir=/home/cpan/bin/perl/bin -Dusedevel -Duse64bitall'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='cc'
    ccflags ='-fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2'
    optimize='-O2'
    cppflags='-fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='11.5.0 20240719 (Red Hat 11.5.0-2)'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib /usr/lib64 /usr/local/lib64
    libs=-lpthread -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -ldl -lm -lcrypt -lutil -lc
    libc=/lib/../lib64/libc.so.6
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version='2.34'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector-strong'


---
@INC for perl 5.41.8:
    /home/cpan/bin/perl/lib/site_perl/5.41.8/x86_64-linux
    /home/cpan/bin/perl/lib/site_perl/5.41.8
    /home/cpan/bin/perl/lib/5.41.8/x86_64-linux
    /home/cpan/bin/perl/lib/5.41.8

---
Environment for perl 5.41.8:
    HOME=/home/cpan
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LC_ALL=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/cpan/bin/perl/bin:/home/cpan/bin:/usr/share/Modules/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash
```

Sample failure report:
http://www.cpantesters.org/cpan/report/83a0af28-cff2-11ef-8dd2-e423a86df0c5

This is a failure during Sereal::Encoder's build-time:
```
Output from '/usr/bin/make':

cp lib/Sereal/Encoder/Constants.pm blib/lib/Sereal/Encoder/Constants.pm
cp lib/Sereal/Encoder.pm blib/lib/Sereal/Encoder.pm
Running Mkbootstrap for Encoder ()
chmod 644 "Encoder.bs"
"/home/cpan/bin/perl/bin/perl5.41.8" -MExtUtils::Command::MM -e 'cp_nonempty' -- Encoder.bs blib/arch/auto/Sereal/Encoder/Encoder.bs 644
"/home/cpan/bin/perl/bin/perl5.41.8" "/home/cpan/bin/perl/lib/5.41.8/ExtUtils/xsubpp"  -typemap '/home/cpan/bin/perl/lib/5.41.8/ExtUtils/typemap' -typemap '/home/cpan/.cpan/build/Sereal-Encoder-5.004-0/typemap'  Encoder.xs > Encoder.xsc
mv Encoder.xsc Encoder.c
cc -c  -I. -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O3   -DVERSION=\"5.004\" -DXS_VERSION=\"5.004\" -fPIC "-I/home/cpan/bin/perl/lib/5.41.8/x86_64-linux/CORE"  -DNDEBUG -DMINIZ_NO_STDIO -DHAVE_ZSTD -DHAVE_BUILTIN_CTZ Encoder.c
cc -c  -I. -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O3   -DVERSION=\"5.004\" -DXS_VERSION=\"5.004\" -fPIC "-I/home/cpan/bin/perl/lib/5.41.8/x86_64-linux/CORE"  -DNDEBUG -DMINIZ_NO_STDIO -DHAVE_ZSTD -DHAVE_BUILTIN_CTZ srl_encoder.c
srl_encoder.c: In function 'srl_dump_hk':
srl_encoder.c:1571:69: error: macro "Perl_bytes_to_utf8" requires 3 arguments, but only 2 given
 1571 |         char* utf8= (char *)Perl_bytes_to_utf8(aTHX_ (U8 *)str, &len);
      |                                                                     ^
In file included from /home/cpan/bin/perl/lib/5.41.8/x86_64-linux/CORE/regexp.h:21,
                 from /home/cpan/bin/perl/lib/5.41.8/x86_64-linux/CORE/perl.h:4522,
                 from srl_encoder.c:8:
/home/cpan/bin/perl/lib/5.41.8/x86_64-linux/CORE/utf8.h:1333:9: note: macro "Perl_bytes_to_utf8" defined here
 1333 | #define Perl_bytes_to_utf8(mTHX, s, lenp)  Perl_bytes_to_utf8_free_me(aTHX_ s, lenp, NULL)
      |         ^~~~~~~~~~~~~~~~~~
srl_encoder.c:1571:29: error: 'Perl_bytes_to_utf8' undeclared (first use in this function); did you mean 'Perl_bytes_from_utf8'?
 1571 |         char* utf8= (char *)Perl_bytes_to_utf8(aTHX_ (U8 *)str, &len);
      |                             ^~~~~~~~~~~~~~~~~~
      |                             Perl_bytes_from_utf8
srl_encoder.c:1571:29: note: each undeclared identifier is reported only once for each function it appears in
make: *** [Makefile:371: srl_encoder.o] Error 1
```
Now bisecting.

Failure output from FreeBSD-13, where `clang17` is the compiler:
```
Output from '/usr/bin/make':

cp lib/Sereal/Encoder/Constants.pm blib/lib/Sereal/Encoder/Constants.pm
cp lib/Sereal/Encoder.pm blib/lib/Sereal/Encoder.pm
Running Mkbootstrap for Encoder ()
chmod 644 "Encoder.bs"
"/usr/home/jkeenan/testing/blead/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- Encoder.bs blib/arch/auto/Sereal/Encoder/Encoder.bs 644
"/usr/home/jkeenan/testing/blead/bin/perl" "/home/jkeenan/testing/blead/lib/perl5/5.41.8/ExtUtils/xsubpp"  -typemap '/home/jkeenan/testing/blead/lib/perl5/5.41.8/ExtUtils/typemap' -typemap '/usr/home/jkeenan/.cpan/build/Sereal-Encoder-5.004-1/typemap'  Encoder.xs > Encoder.xsc
mv Encoder.xsc Encoder.c
cc -c  -I.  -DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -DNO_POSIX_2008_LOCALE -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -O3 -Wno-compound-token-split-by-macro    -DVERSION=\"5.004\"  -DXS_VERSION=\"5.004\" -DPIC -fPIC "-I/home/jkeenan/testing/blead/lib/perl5/5.41.8/amd64-freebsd-thread-multi/CORE"  -DNDEBUG -DMINIZ_NO_STDIO -DHAVE_ZSTD -DHAVE_BUILTIN_CTZ Encoder.c
cc -c  -I.  -DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -DNO_POSIX_2008_LOCALE -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -O3 -Wno-compound-token-split-by-macro    -DVERSION=\"5.004\"  -DXS_VERSION=\"5.004\" -DPIC -fPIC "-I/home/jkeenan/testing/blead/lib/perl5/5.41.8/amd64-freebsd-thread-multi/CORE"  -DNDEBUG -DMINIZ_NO_STDIO -DHAVE_ZSTD -DHAVE_BUILTIN_CTZ srl_encoder.c
srl_encoder.c:1571:69: error: too few arguments provided to function-like macro invocation
 1571 |         char* utf8= (char *)Perl_bytes_to_utf8(aTHX_ (U8 *)str, &len);
      |                                                                     ^
/home/jkeenan/testing/blead/lib/perl5/5.41.8/amd64-freebsd-thread-multi/CORE/utf8.h:1333:9: note: macro 'Perl_bytes_to_utf8' defined here
 1333 | #define Perl_bytes_to_utf8(mTHX, s, lenp)  Perl_bytes_to_utf8_free_me(aTHX_ s, lenp, NULL)
      |         ^
srl_encoder.c:1571:29: error: use of undeclared identifier 'Perl_bytes_to_utf8'
 1571 |         char* utf8= (char *)Perl_bytes_to_utf8(aTHX_ (U8 *)str, &len);
      |                             ^
2 errors generated.
*** Error code 1

Stop.
make: stopped in /usr/home/jkeenan/.cpan/build/Sereal-Encoder-5.004-1
```

--- a/embed.fnc
+++ b/embed.fnc
@@ -794,12 +794,12 @@ Adp	|int	|bytes_cmp_utf8 |NN const U8 *b 			\
 Adp	|U8 *	|bytes_from_utf8|NN const U8 *s 			\
 				|NN STRLEN *lenp			\
 				|NN bool *is_utf8p
-Admp	|U8 *	|bytes_to_utf8	|NN const U8 *s 			\
+Adip	|U8 *	|bytes_to_utf8	|NN const U8 *s 			\
 				|NN STRLEN *lenp
 Adp	|U8 *	|bytes_to_utf8_free_me					\
 				|NN const U8 *s 			\
 				|NN STRLEN *lenp			\
-				|NULLOK const U8 **free_me
+				|NULLOK void **free_me
 AOdp	|SSize_t|call_argv	|NN const char *sub_name		\
 				|I32 flags				\
 				|NN char **argv

--- a/embed.h
+++ b/embed.h
@@ -155,7 +155,7 @@
 # define block_start(a)                         Perl_block_start(aTHX_ a)
 # define bytes_cmp_utf8(a,b,c,d)                Perl_bytes_cmp_utf8(aTHX_ a,b,c,d)
 # define bytes_from_utf8(a,b,c)                 Perl_bytes_from_utf8(aTHX_ a,b,c)
-# define bytes_to_utf8(a,b)                     Perl_bytes_to_utf8(aTHX,a,b)
+# define bytes_to_utf8(a,b)                     Perl_bytes_to_utf8(aTHX_ a,b)
 # define bytes_to_utf8_free_me(a,b,c)           Perl_bytes_to_utf8_free_me(aTHX_ a,b,c)
 # define c9strict_utf8_to_uv                    Perl_c9strict_utf8_to_uv
 # define call_argv(a,b,c)                       Perl_call_argv(aTHX_ a,b,c)

--- a/inline.h
+++ b/inline.h
@@ -1132,7 +1132,6 @@ in lvalue context.
 =cut
 */
 
-
 PERL_STATIC_INLINE bool
 Perl_rpp_is_lone(pTHX_ SV *sv)
 {
@@ -1229,6 +1228,12 @@ Perl_append_utf8_from_native_byte(const U8 byte, U8** dest)
         *((*dest)++) = UTF8_EIGHT_BIT_HI(byte);
         *((*dest)++) = UTF8_EIGHT_BIT_LO(byte);
     }
+}
+
+PERL_STATIC_INLINE U8 *
+Perl_bytes_to_utf8(pTHX_ const U8 *s, STRLEN *lenp)
+{
+    return bytes_to_utf8_free_me(s, lenp, NULL);
 }
 
 /*

--- a/proto.h
+++ b/proto.h
@@ -398,11 +398,8 @@ Perl_bytes_from_utf8(pTHX_ const U8 *s, STRLEN *lenp, bool *is_utf8p);
 #define PERL_ARGS_ASSERT_BYTES_FROM_UTF8        \
         assert(s); assert(lenp); assert(is_utf8p)
 
-/* PERL_CALLCONV U8 *
-Perl_bytes_to_utf8(pTHX_ const U8 *s, STRLEN *lenp); */
-
 PERL_CALLCONV U8 *
-Perl_bytes_to_utf8_free_me(pTHX_ const U8 *s, STRLEN *lenp, const U8 **free_me);
+Perl_bytes_to_utf8_free_me(pTHX_ const U8 *s, STRLEN *lenp, void **free_me);
 #define PERL_ARGS_ASSERT_BYTES_TO_UTF8_FREE_ME  \
         assert(s); assert(lenp)
 
@@ -9650,6 +9647,11 @@ PERL_STATIC_INLINE SV **
 Perl_av_store_simple(pTHX_ AV *av, SSize_t key, SV *val);
 # define PERL_ARGS_ASSERT_AV_STORE_SIMPLE       \
         assert(av); assert(SvTYPE(av) == SVt_PVAV)
+
+PERL_STATIC_INLINE U8 *
+Perl_bytes_to_utf8(pTHX_ const U8 *s, STRLEN *lenp);
+# define PERL_ARGS_ASSERT_BYTES_TO_UTF8         \
+        assert(s); assert(lenp)
 
 PERL_STATIC_INLINE void
 Perl_clear_defarray_simple(pTHX_ AV *av);

--- a/utf8.c
+++ b/utf8.c
@@ -3269,7 +3269,7 @@ But when it is a non-NULL pointer, C<bytes_to_utf8_free_me> stores into it
 either NULL if no memory was allocated; or a pointer to that new memory.  This
 allows the following convenient paradigm:
 
- U8 * free_me;
+ void * free_me;
  U8 converted = bytes_to_utf8_free_me(string, &len, &free_me);
 
  ...
@@ -3292,7 +3292,7 @@ EBCDIC), see L</sv_recode_to_utf8>().
 
 U8*
 Perl_bytes_to_utf8_free_me(pTHX_ const U8 *s, Size_t *lenp,
-                                 const U8 ** free_me_ptr)
+                                 void ** free_me_ptr)
 {
     PERL_ARGS_ASSERT_BYTES_TO_UTF8_FREE_ME;
     PERL_UNUSED_CONTEXT;

--- a/utf8.h
+++ b/utf8.h
@@ -1330,7 +1330,6 @@ point's representation.
 
 #define Perl_is_utf8_char_buf(buf, buf_end) isUTF8_CHAR(buf, buf_end)
 
-#define Perl_bytes_to_utf8(mTHX, s, lenp)  Perl_bytes_to_utf8_free_me(aTHX_ s, lenp, NULL)
 typedef enum {
     PL_utf8_to_bytes_overwrite = 0,
     PL_utf8_to_bytes_new_memory,

